csrc/runtime/compiled_kernel.h

```diff
@@ -0,0 +1,305 @@
+//! string. This util provides a path to test runtime code, i.e. the resource
+//! strings.
+class RtcKernel : public NonCopyable {
+ public:
+  NVF_API void compile(
+      const std::string& code,
+      const std::string& name,
+      bool structured,
+      PrimDataType index_type,
+      int64_t device_index = 0);
+
+  //! Internal tests only. Runs the compiled CUDA kernel. Returns elapsed
+  //! milliseconds.
+  NVF_API float run(
+      const LaunchParams& launch_params,
+      const std::vector<at::Tensor>& args,
+      PrimDataType indextype);
+
+ private:
+  std::unique_ptr<executor_utils::CudaExecutable> compiled_kernel_;
+  int64_t device_index_;
+};
+
+//! Class for compilation logic through nvRTC. It shouldn't hold any logic
+//! associated with how to run a kernel, but how to compile it. It should also
+//! contain any information about the kernel itself.
+class CompiledKernel : public NonCopyable {
+ public:
+  // NVF_API was added for nvfuser_extension. See examples/sinh_extension.
+  CompiledKernel() = delete;
+
+  NVF_API CompiledKernel(
+      Fusion* fusion,
+      CompileParams compile_params,
+      c10::Device device,
+      SchedulerType scheduler_type,
+      int64_t fusion_id,
+      int64_t concrete_id,
+      int64_t runtime_id,
+      int64_t group_id,
+      const std::vector<std::function<void(GpuLower*)>>& pre_lowering_hooks,
+      const std::vector<std::function<void(kir::Kernel*)>>&
+          post_lowering_hooks);
+
+  NVF_API CompiledKernel(
+      Fusion* fusion,
+      CompileParams compile_params,
+      c10::Device device,
+      SchedulerType scheduler_type = SchedulerType::None,
+      int64_t fusion_id = 0,
+      int64_t concrete_id = 0,
+      int64_t runtime_id = 0,
+      int64_t group_id = 0);
+
+  CompiledKernel(const CompiledKernel& other) = delete;
+  CompiledKernel& operator=(const CompiledKernel& other) = delete;
+  CompiledKernel(CompiledKernel&& other) noexcept = delete;
+  CompiledKernel& operator=(CompiledKernel&& other) noexcept = delete;
+
+  //! To compile a fusion with the 32-bit index type, CompileParams
+  //! must be passed in. There used to be an index type associated
+  //! with KernelArgumentHolder, but it is no longer the case.
+  NVF_API void compile(int64_t block_size);
+
+  // Function to query whether compilation was attempted for a `CompiledKernel`
+  bool isCompiled() const {
+    return lowered_ != nullptr;
+  };
+
+  // function to query whether a `CompiledKernel` has a compiled kernel to
+  // execute
+  bool hasCompiledKernel() const {
+    if (compiled_kernel_ != nullptr) {
+      NVF_ERROR(compiled_kernel_->function != nullptr);
+    }
+    return validKernelId() && lowered_ && compiled_kernel_ != nullptr;
+  };
+
+  using ExecutorCompileTimeInfoCache =
+      executor_utils::caching::ExecutorCompileTimeInfoCache;
+
+  kir::Kernel* kernel() const {
+    NVF_ERROR(lowered_);
+    return lowered_->kernel();
+  }
+
+  Fusion* fusion() const {
```

Comment:
I believe this isn't any more useful than kernel() so I removed this in https://github.com/NVIDIA/Fuser/pull/3725.

Reply:
Removing.